### PR TITLE
Threejs-CCW

### PR DESCRIPTION
**Pull request by xiaochen004hao**
# **共创世界-3D扩展！**
### [示例demo](https://www.ccw.site/gandi/project/65fecfcfdf30e51731df0e25)
---

##### _PR一定要过啊！！！这个扩展我付出了很多心血（因为不可描述的原因，我只能周六上线）_
# 🫶🫶🫶

### CODE REVIEW

**Comment by xiaochen004hao** [comment]
# **HCN可以把这个扩展放到”开发中的扩展“**

File: ThreejsCCW/ThreejsCCW.js [hunk 1]
```diff
@@ -0,0 +1,1288 @@
...
+        // 环境光
+        this.ambient_light = null;
+
+        // 光
+        this.lights = new Object();
+        // 物体
+        this.objects = new Object();
+
+        // 原CCW显示canvas
+        this._ccw = document.getElementsByClassName('gandi_stage_stage_1fD7k')[0].getElementsByTagName("canvas")[0];
```
Comment (Nightre) [inline ThreejsCCW/ThreejsCCW.js:53]: 可以用 runtime.renderer.canvas

**Comment by fath11** [comment]
1. You should make the custom canvas automatically resize, look at BBCode's source
2. The demo project isn't shared? its says no access and i cant get the extension to work 
image

**Comment by fath11** [comment]
What I am worrying about is support for Quake and GKamera. Looking at how the extension works, GKamera won't be a problem, but the Quake extension might be a different story.

Maybe this extension might need its own "load shader" block. If you look at #204 it's possible to get files from the Code Editor. Since i myself have been experimenting with Three.js for the past 3 months, i can create this feature if i have the time but don't expect me to work on it. Tho it will be great if you can

File: ThreejsCCW/ThreejsCCW.js [hunk 2]
```diff
@@ -0,0 +1,1288 @@
+// 依赖库
+import * as THREE from "https://unpkg.com/three@0.162.0/build/three.module.js";
+
+// 插件
+// 这里为该扩展目录下的三个js文件
+import {OBJLoader} from "https://livefile.xesimg.com/programme/python_assets/5bcc04d1d301d7a3e096fa6b8708af7f.js";
```
Comment (FurryR) [inline ThreejsCCW/ThreejsCCW.js:6]: 先有学而思后有天。请勿使用学而思临时文件。可以参考 lpp 那样，配置一个脚手架/打包器。

File: ThreejsCCW/ThreejsCCW.js [hunk 3]
```diff
@@ -0,0 +1,1288 @@
...
+import {OrbitControls} from 'https://livefile.xesimg.com/programme/python_assets/22b053c1486c0311a7e493e9419cebf7.js';
+import {MTLLoader} from 'https://livefile.xesimg.com/programme/python_assets/bfdf94e29d1b77c18dbfc5aa778e6df7.js';
+// import {OBJLoader} from "https://cdn.jsdelivr.net/gh/MoreBugOfDog/cdn-file@TBS/file/OBJLoader.js";
+// import {OrbitControls} from 'https://cdn.jsdelivr.net/gh/MoreBugOfDog/cdn-file@TBS/file/OrbitControls.js';
+// import {MTLLoader} from 'https://cdn.jsdelivr.net/gh/MoreBugOfDog/cdn-file@TBS/file/MTLLoader.js';
+import WebGL from 'https://unpkg.com/three@0.162.0/examples/jsm/capabilities/WebGL.js';
+
+console.log(THREE);
```
Comment (FurryR) [inline ThreejsCCW/ThreejsCCW.js:14]:
```suggestion
```

调试信息不应当出现在送审的版本。

File: ThreejsCCW/ThreejsCCW.js [hunk 4]
```diff
@@ -0,0 +1,1288 @@
...
+        this.near = null;
+        this.far = null;
+        this.camera = null;
+        this.controls = null;
+
+        // 环境光
+        this.ambient_light = null;
+
+        // 光
+        this.lights = new Object();
```
Comment (FurryR) [inline ThreejsCCW/ThreejsCCW.js:48]:
```suggestion
        this.lights = {};
```

File: ThreejsCCW/ThreejsCCW.js [hunk 5]
```diff
@@ -0,0 +1,1288 @@
...
+     * @param {number} args.color
+     * @param {number} args.sizey
+     * @param {number} args.sizex
+     * @param {string} args.Anti_Aliasing
+     */
+    init(args) {
+        // this._ccw.style.display = "none";  // 隐藏原CCW显示canvas
+
+        // 创建threejs显示canvas
+        if (document.getElementsByClassName('gandi_stage_stage_1fD7k')[0].getElementsByClassName("ThreejsCCW").length == 0) {
```
Comment (FurryR) [inline ThreejsCCW/ThreejsCCW.js:787]: 硬编码这种带随机数的 id 会在下一次更新时立即失效。考虑改用 runtime.renderer.canvas。

File: ThreejsCCW/ThreejsCCW.js [hunk 6]
```diff
@@ -0,0 +1,1288 @@
...
+
+    /**
+     * 初始化
+     * @param {number} args.color
+     * @param {number} args.sizey
+     * @param {number} args.sizex
+     * @param {string} args.Anti_Aliasing
+     */
+    init(args) {
+        // this._ccw.style.display = "none";  // 隐藏原CCW显示canvas
```
Comment (FurryR) [inline ThreejsCCW/ThreejsCCW.js:784]: 为什么不能原版和 ThreeJS 共存呢？

**Review by FurryR — CHANGES_REQUESTED** [review]
勇气可嘉，技术太菜。隔壁我记得有人做过兼容 Scratch 原版的 Three 了。

**Comment by xiaochen004hao** [comment]
我只能周六用电脑。。。

在 2024-03-24 20:08:54，"凌" ***@***.***> 写道：


@FurryR commented on this pull request.

In ThreejsCCW/ThreejsCCW.js:

> @@ -0,0 +1,1288 @@
+// 依赖库
+import * as THREE from ***@***.***/build/three.module.js";
+
+// 插件
+// 这里为该扩展目录下的三个js文件
+import {OBJLoader} from "https://livefile.xesimg.com/programme/python_assets/5bcc04d1d301d7a3e096fa6b8708af7f.js";
+import {OrbitControls} from 'https://livefile.xesimg.com/programme/python_assets/22b053c1486c0311a7e493e9419cebf7.js';
+import {MTLLoader} from 'https://livefile.xesimg.com/programme/python_assets/bfdf94e29d1b77c18dbfc5aa778e6df7.js';
+// import {OBJLoader} from ***@***.***/file/OBJLoader.js";
+// import {OrbitControls} from ***@***.***/file/OrbitControls.js';
+// import {MTLLoader} from ***@***.***/file/MTLLoader.js';
+import WebGL from ***@***.***/examples/jsm/capabilities/WebGL.js';
+
+console.log(THREE);

⬇️ Suggested change
-console.log(THREE);


调试信息不应当出现在送审的版本。

**Comment by xiaochen004hao** [comment]
To be frank, who are you? I can only use the computer on Saturdays... Strive to revise all the suggestions completely

在 2024-03-24 17:59:02，fath11 ***@***.***> 写道：


What I am worrying about is support for Quake and GKamera. Looking at how the extension works, GKamera won't be a problem, but the Quake extension might be a different story.

Maybe this extension might need its own "load shader" block. If you look at #204 it's possible to get files from the Code Editor. Since i myself have been experimenting with Three.js for the past 3 months, i can create this feature if i have the time but don't expect me to work on it. Tho it will be great if you can

**Comment by xiaochen004hao** [comment]
因为我js dom很菜，菜就多练的那种

在 2024-03-24 20:12:10，"凌" ***@***.***> 写道：


@FurryR commented on this pull request.

In ThreejsCCW/ThreejsCCW.js:

> +            })
+        }
+     
+        this.scene.remove(model);
+
+    }
+
+    /**
+     * 初始化
+     * @param {number} args.color
+     * @param {number} args.sizey
+     * @param {number} args.sizex
+     * @param {string} args.Anti_Aliasing
+     */
+    init(args) {
+        // this._ccw.style.display = "none";  // 隐藏原CCW显示canvas


为什么不能原版和 ThreeJS 共存呢？

**Comment by xiaochen004hao** [comment]
那么，objects是不是也得这样搞

在 2024-03-24 20:09:39，"凌" ***@***.***> 写道：


@FurryR commented on this pull request.

In ThreejsCCW/ThreejsCCW.js:

> +        // 场景
+        this.scene = null;
+
+        // 相机配置
+        this.fov = null;
+        this.aspect = null;
+        this.near = null;
+        this.far = null;
+        this.camera = null;
+        this.controls = null;
+
+        // 环境光
+        this.ambient_light = null;
+
+        // 光
+        this.lights = new Object();

⬇️ Suggested change
-        this.lights = new Object();
+        this.lights = {};

**Comment by xiaochen004hao** [comment]
> 勇气可嘉，技术太菜。隔壁我记得有人做过兼容 Scratch 原版的 Three 了。

神奇，在github发的评论在我的网易邮箱写能看到唉

**Comment by fath11** [comment]
> 勇气可嘉，技术太菜。隔壁我记得有人做过兼容 Scratch 原版的 Three 了。Courage is commendable, and technology is too good. Next door I remember someone making a Three that was compatible with the original Scratch version.

The extension uses skins. I am also trying to make a 3D extension that will coexist with Scratch without the use of skins but it requires me to learn more about WebGL which is very time-consuming as I have 0 prior knowledge about it. I am getting closer tho, the only thing slowing me down is school.

I am actually surprised someone made this PR.

**Comment by xiaochen004hao** [comment]
> 1. You should make the custom canvas automatically resize, look at BBCode's source
> 2. The demo project isn't shared? its says no access and i cant get the extension to work
>    image
> 3. Perhaps you should consider changing the extension's name to avoid confusion with Three

可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）

**Comment by xiaochen004hao** [comment]
> 勇气可嘉，技术太菜。隔壁我记得有人做过兼容 Scratch 原版的 Three 了。
还要，我几乎没学过js，当c++ + Python来写的🙂

**Comment by kukemc** [comment]
> > 1. 您应该让自定义画布自动调整大小，查看 BBCode 的源代码
> > 2. 演示项目没有共享？它说没有访问权限，我无法让扩展程序工作
> >    image
> > 3. 也许您应该考虑更改扩展名以避免与 Three 混淆
> 
> 可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）

你应该将扩展挂到cdn上然后直接发链接

**Comment by xiaochen004hao** [comment]
> > > 1. 您应该让自定义画布自动调整大小，查看 BBCode 的源代码
> > > 2. 演示项目没有共享？它说没有访问权限，我无法让扩展程序工作
> > >    image
> > > 3. 也许您应该考虑更改扩展名以避免与 Three 混淆
> > 
> > 
> > 可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）
> 
> 你应该将扩展挂到cdn上然后直接发链接

你是，酷可！

**Comment by xiaochen004hao** [comment]
> > > 1. 您应该让自定义画布自动调整大小，查看 BBCode 的源代码
> > > 2. 演示项目没有共享？它说没有访问权限，我无法让扩展程序工作
> > >    image
> > > 3. 也许您应该考虑更改扩展名以避免与 Three 混淆
> > 
> > 
> > 可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）
> 
> 你应该将扩展挂到cdn上然后直接发扩展url

还有

**Comment by xiaochen004hao** [comment]
> > > 1. 您应该让自定义画布自动调整大小，查看 BBCode 的源代码
> > > 2. 演示项目没有共享？它说没有访问权限，我无法让扩展程序工作
> > >    image
> > > 3. 也许您应该考虑更改扩展名以避免与 Three 混淆
> > 
> > 
> > 可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）
> 
> 你应该将扩展挂到cdn上然后直接发扩展url

唉 我刚才按了什么

**Comment by MoreBugOfDog** [comment]
> > > > 1. 您应该让自定义画布自动调整大小，查看 BBCode 的源代码
> > > > 2. 演示项目没有共享？它说没有访问权限，我无法让扩展程序工作
> > > >    图像
> > > > 3. 也许您应该考虑更改扩展名以避免与 Three 混淆
> > > 
> > > 
> > > 可是，装了自定义扩展的作品发不出来，难道我要发观察者邀请链接（应该是的）
> > 
> > 
> > 你应该将扩展挂到cdn上，然后直接发扩展url
> 
> 唉 我刚才按了什么

你**关闭**了这个PR，也就是说你取消了审核。。
如果需要审核，请**再次发起PR**

**Comment by xiaochen004hao** [comment]
考古，其实这里也是我第一次使用Github的地方

File: ThreejsCCW/ThreejsCCW.js (hunk 2 above)
Comment (hwxqyworld) [reply ThreejsCCW/ThreejsCCW.js:6]: > 先有学而思后有天。不要使用学而思临时文件。可以参考lpp那样，配置一个脚手架/备份器。

捉

**Pull request closed without merging**